beacon-chain/slasher/helpers.go

```diff
@@ -0,0 +1,140 @@
+package slasher
+
+import (
+	"strconv"
+
+	types "github.com/prysmaticlabs/eth2-types"
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	slashertypes "github.com/prysmaticlabs/prysm/beacon-chain/slasher/types"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/sliceutil"
+	"github.com/sirupsen/logrus"
+)
+
+// Group a list of attestations into batches by validator chunk index.
+// This way, we can detect on the batch of attestations for each validator chunk index
+// concurrently, and also allowing us to effectively use a single 2D chunk
+// for slashing detection through this logical grouping.
+func (s *Service) groupByValidatorChunkIndex(
+	attestations []*slashertypes.IndexedAttestationWrapper,
+) map[uint64][]*slashertypes.IndexedAttestationWrapper {
+	groupedAttestations := make(map[uint64][]*slashertypes.IndexedAttestationWrapper)
+	for _, att := range attestations {
+		validatorChunkIndices := make(map[uint64]bool)
+		for _, validatorIdx := range att.IndexedAttestation.AttestingIndices {
+			validatorChunkIndex := s.params.validatorChunkIndex(types.ValidatorIndex(validatorIdx))
+			validatorChunkIndices[validatorChunkIndex] = true
+		}
+		for validatorChunkIndex := range validatorChunkIndices {
+			groupedAttestations[validatorChunkIndex] = append(
+				groupedAttestations[validatorChunkIndex],
+				att,
+			)
+		}
+	}
+	return groupedAttestations
+}
+
+// Group attestations by the chunk index their source epoch corresponds to.
+func (s *Service) groupByChunkIndex(
+	attestations []*slashertypes.IndexedAttestationWrapper,
+) map[uint64][]*slashertypes.IndexedAttestationWrapper {
+	attestationsByChunkIndex := make(map[uint64][]*slashertypes.IndexedAttestationWrapper)
+	for _, att := range attestations {
+		chunkIdx := s.params.chunkIndex(att.IndexedAttestation.Data.Source.Epoch)
+		attestationsByChunkIndex[chunkIdx] = append(attestationsByChunkIndex[chunkIdx], att)
+	}
+	return attestationsByChunkIndex
+}
+
+// This function returns a list of valid attestations, a list of attestations that are
+// valid in the future, and the number of attestations dropped.
+func (s *Service) filterAttestations(
+	atts []*slashertypes.IndexedAttestationWrapper, currentEpoch types.Epoch,
+) (valid, validInFuture []*slashertypes.IndexedAttestationWrapper, numDropped int) {
+	valid = make([]*slashertypes.IndexedAttestationWrapper, 0, len(atts))
+	validInFuture = make([]*slashertypes.IndexedAttestationWrapper, 0, len(atts))
+
+	for _, attWrapper := range atts {
+		if attWrapper == nil || !validateAttestationIntegrity(attWrapper.IndexedAttestation) {
+			numDropped++
+			continue
+		}
+
+		// If an attestation's source is epoch is older than the max history length
+		// we keep track of for slashing detection, we drop it.
+		if attWrapper.IndexedAttestation.Data.Source.Epoch+types.Epoch(s.params.historyLength) <= currentEpoch {
```

Comment:
```suggestion
		if attWrapper.IndexedAttestation.Data.Source.Epoch+s.params.historyLength <= currentEpoch {
```
`historyLength` is of `Epoch` type.